### PR TITLE
JIT: Optimize data flow used in assertion prop/CSE

### PR DESCRIPTION
**Pull request by jakobbotsch**
The data flow used by assertion prop and CSE utilize "all-succs" which includes handler successors. However, in reality neither analysis needs full treatment of handlers; we do not CSE variables that are live-in to handlers, and assertion prop never kills assertions that would need to be propagated to handlers.

Additionally, the data flow framework used conflates end-of-block propagation of facts with reachability of the handler. If no facts changed at the ends of the preds of the handler, then the handler is not visited. This is despite the fact that the handler in the general case is also reachable with facts from the beginning of every enclosed basic block (or more generally with facts that were invalidated in the middle of enclosed basic blocks).

This ends up working out today only because of the fake successor edges we have from preds of the try to the handler, in addition to the restrictions on the data flows described above. Since I'm removing these fake edges, we need a more explicit solution. The change here has such a solution, by making it more explicit what exactly is needed in the data flow for CSE and AP here: they only need to consider regular successors, with the added catch that they need to consider reachability of handlers once we see the corresponding 'try' being reachable.

Prerequisite to #94672.

No diffs are expected.

### CODE REVIEW

File: src/coreclr/jit/assertionprop.cpp [hunk 1]
```diff
@@ -5498,11 +5505,8 @@ class AssertionPropFlowCallback
             Compiler::optDumpAssertionIndices("in -> ", block->bbAssertionIn, "; ");
             JITDUMP("firstTryBlock " FMT_BB " ", firstTryBlock->bbNum);
             Compiler::optDumpAssertionIndices("in -> ", firstTryBlock->bbAssertionIn, "; ");
-            JITDUMP("lastTryBlock " FMT_BB " ", lastTryBlock->bbNum);
-            Compiler::optDumpAssertionIndices("out -> ", lastTryBlock->bbAssertionOut, "\n");
         }
         BitVecOps::IntersectionD(apTraits, block->bbAssertionIn, firstTryBlock->bbAssertionIn);
-        BitVecOps::IntersectionD(apTraits, block->bbAssertionIn, lastTryBlock->bbAssertionOut);
```
Comment (jakobbotsch) [inline src/coreclr/jit/assertionprop.cpp:5505]: This was pretty meaningless -- the lexically last block of a try region shouldn't be sufficient to guarantee any form of correctness.

**Comment by jakobbotsch** [comment]
cc @dotnet/jit-contrib PTAL @AndyAyersMS 

[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=469223&view=ms.vss-build-web.run-extensions-tab) -- only a few on x86, that I investigated above. Other than that just some nice throughput improvements.

File: src/coreclr/jit/assertionprop.cpp [hunk 2]
```diff
@@ -5488,8 +5488,15 @@ class AssertionPropFlowCallback
...
-    //   We can jump to the handler from any instruction in the try region.
-    //   It means we can propagate only assertions that are valid for the whole try region.
+    //   We can jump to the handler from any instruction in the try region. It
+    //   means we can propagate only assertions that are valid for the whole
+    //   try region.
+    //
+    //   It suffices to intersect with only the head 'try' block's assertions,
+    //   since that block dominates all other blocks in the try and we never
+    //   kill assertions in global AP. Note that if we did kill assertions we
+    //   would need to be more careful about our mid-block handling when in a
```
Comment (SingleAccretion) [inline src/coreclr/jit/assertionprop.cpp:5498]: >    Note that if we did kill assertions...

In my opinion, it is confusing to talk about the possibility of "killing assertions" in global AP. How would that work? It seems impossible almost as if by definition of the algorithm.

File: src/coreclr/jit/dataflow.h [hunk 3]
```diff
@@ -58,19 +58,47 @@ void DataFlow::ForwardAnalysis(TCallback& callback)
         }
         else
         {
-            FlowEdge* preds = m_pCompiler->BlockPredsWithEH(block);
-            for (FlowEdge* pred = preds; pred; pred = pred->getNextPredEdge())
+            for (FlowEdge* pred = block->bbPreds; pred; pred = pred->getNextPredEdge())
```
Comment (SingleAccretion) [inline src/coreclr/jit/dataflow.h:61]: Nit: can this loop use the `PredEdges` iterator?

File: src/coreclr/jit/dataflow.h [hunk 4]
```diff
@@ -58,19 +58,47 @@ void DataFlow::ForwardAnalysis(TCallback& callback)
...
+            // that block dominates all other blocks in the 'try'. That happens
+            // implicitly below.
+            //
+            block->VisitRegularSuccs(m_pCompiler, [&worklist](BasicBlock* succ) {
                 worklist.insert(worklist.end(), succ);
                 return BasicBlockVisit::Continue;
             });
         }
+
+        if (m_pCompiler->bbIsTryBeg(block))
```
Comment (SingleAccretion) [inline src/coreclr/jit/dataflow.h:86]: Why do we not need to consider the exceptional successors of filters here?

(Not saying it should)

Comment (jakobbotsch) [reply src/coreclr/jit/dataflow.h:86]: Do you mean the second pass EH successors?

File: src/coreclr/jit/assertionprop.cpp (hunk 2 above)
Comment (jakobbotsch) [reply src/coreclr/jit/assertionprop.cpp:5498]: Do you mean because we use SSA/VN? I mean if facts could become false in other ways than due to control flow joins. We would need to propagate that to the handlers.

File: src/coreclr/jit/dataflow.h (hunk 4 above)
Comment (SingleAccretion) [reply src/coreclr/jit/dataflow.h:86]: Yes.

File: src/coreclr/jit/assertionprop.cpp (hunk 2 above)
Comment (SingleAccretion) [reply src/coreclr/jit/assertionprop.cpp:5498]: > Do you mean because we use SSA/VN?

Right. VNs are "immutable", so if a a fact about a VN dominates a subgraph, it will always be true on the entirety of that subgraph. I understand the comment, I would just request to reword to not mention AP along the lines of "no clients need the concept of 'kill's".

File: src/coreclr/jit/dataflow.h (hunk 4 above)
Comment (jakobbotsch) [reply src/coreclr/jit/dataflow.h:86]: I think that because the filter and its enclosed handlers are all dominated by a same try entry there is no need to consider the filter as a pred of those handlers.

File: src/coreclr/jit/assertionprop.cpp (hunk 2 above)
Comment (jakobbotsch) [reply src/coreclr/jit/assertionprop.cpp:5498]: Sure, I can reword it tomorrow.

File: src/coreclr/jit/dataflow.h (hunk 3 above)
Comment (jakobbotsch) [reply src/coreclr/jit/dataflow.h:61]: Yeah, let me simplify this tomorrow.

File: src/coreclr/jit/assertionprop.cpp (hunk 2 above)
Comment (jakobbotsch) [reply src/coreclr/jit/assertionprop.cpp]: Do you think the wording looks better now?

**Comment by jakobbotsch** [comment]
Ping @AndyAyersMS... when you have time.

**Review by AndyAyersMS — APPROVED** [review]
